[LV] Introduce the EVLIVSimplify Pass for EVL-vectorized loops

mshockwave opened the pull request:
When we enable EVL-based loop vectorization w/ predicated tail-folding, each vectorized loop has effectively two induction variables: one calculates the step using (VF x vscale) and the other one increases the IV by values returned from experiment.get.vector.length. The former, also known as canonical IV, is more favorable for analyses as it's "countable" in the sense of SCEV; the latter (EVL-based IV), however, is more favorable to codegen, at least for those that support scalable vectors like AArch64 SVE and RISC-V.

The idea is that we use canonical IV all the way until the beginning of codegen pipeline, where we replace it with EVL-based IV using EVLIVSimplify introduced here. Such that we can have the best from both worlds.

This Pass is enabled by default in RISC-V. However, since we haven't really vectorize loops with predicate tail-folding, this Pass is no-op at this moment.

That said, I have validate the correctness of this Pass by enable EVL-based LV + predicated tail-folding
(i.e. `-force-tail-folding-style=data-with-evl`
`-prefer-predicate-over-epilogue=predicate-dont-vectorize`) and run on SPEC2006INT and SPEC2017 intrate w/ test workload.
With this setup and `-mcpu=sifive-x280`, I'd observed a notable 1.8% improvement on dynamic instruction counts on 525.x264_r. Aside from that, rest of benchmarks only have negligible (dynamic IC) differences which are less than 1%.

mshockwave committed:
[LV] Introduce the EVLIVSimplify Pass for EVL-vectorized loops

When we enable EVL-based loop vectorization w/ predicated tail-folding,
each vectorized loop has effectively two induction variables: one
calculates the step using (VF x vscale) and the other one increases the
IV by values returned from experiment.get.vector.length. The former,
also known as canonical IV, is more favorable for analyses as it's "countable"
in the sense of SCEV; the latter (EVL-based IV), however, is more favorable to
codegen, at least for those that support scalable vectors like AArch64 SVE and
RISC-V.

The idea is that we use canonical IV all the way until the beginning of codegen
pipeline, where we replace it with EVL-based IV using EVLIVSimplify
introduced here. Such that we can have the best from both worlds.

This Pass is enabled by default in RISC-V. However, since we haven't
really vectorize loops with predicate tail-folding, this Pass is
no-op at this moment.

That said, I have validate the correctness of this Pass by enable
EVL-based LV + predicated tail-folding
(i.e. -force-tail-folding-style=data-with-evl
-prefer-predicate-over-epilogue=predicate-dont-vectorize) and run on
SPEC2006INT and SPEC2017 intrate w/ test workload.

alexey-bataev commented on llvm/lib/Transforms/Vectorize/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,253 @@
+
+  // If not, see if the vscale_range of the parent function is a fixed value,
+  // which makes the step value to be replaced by a constant.
+  if (isa<SCEVConstant>(Step) && F.hasFnAttribute(Attribute::VScaleRange)) {
+    APInt V = cast<SCEVConstant>(Step)->getAPInt().abs();
```
Use dyn_cast instead of isa/cast

mshockwave replied:
Done.

alexey-bataev commented on llvm/test/Transforms/LoopVectorize/RISCV/evl-iv-simplify.ll:
```diff
@@ -0,0 +1,282 @@
+; RUN: opt -S -mtriple=riscv64 -mattr='+v' --passes='loop(evl-iv-simplify)' < %s | FileCheck %s
+; RUN: opt -S -mtriple=riscv64 -mattr='+v' --passes='loop(evl-iv-simplify),function(simplifycfg,dce)' < %s | FileCheck %s --check-prefix=LOOP-DEL
+
+define void @simple(ptr noalias %a, ptr noalias %b, ptr noalias %c, i64 %N) {
```
Would be good to simplify the test as much as possible

mshockwave replied:
I've tried to simplify it by only using only a single vp.load rather than two.

mshockwave committed:
Address review comments

And simplify the test cases.

arcbbb commented on llvm/lib/Transforms/Vectorize/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,254 @@
+
+  Value *EVLIndex = nullptr;
+  Value *RemVL = nullptr, *AVL = nullptr;
+  auto IntrinsicMatch = m_Intrinsic<Intrinsic::experimental_get_vector_length>(
+      m_Value(RemVL), m_SpecificInt(*VF),
+      /*Scalable=*/m_SpecificInt(1));
+  for (auto &PN : BB->phis()) {
```
Since there is only one phi node in `match(RemVL, m_Sub(m_Value(AVL), m_Specific(&PN))))`,
can we check the node exclusively?

arcbbb replied:
Ah I see. We don't know which `get.vector.length` to match at the place.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/evl-iv-simplify.ll:
```diff
@@ -0,0 +1,264 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -S -mtriple=riscv64 -mattr='+v' --passes='loop(evl-iv-simplify)' < %s | FileCheck %s
+; RUN: opt -S -mtriple=riscv64 -mattr='+v' --passes='loop(evl-iv-simplify),function(simplifycfg,dce)' < %s | FileCheck %s --check-prefix=LOOP-DEL
```
`evl-iv-simplify` is a separate pass and should have separate tests, independent of LV tests.

fhahn reviewed (commented):
as this is a backend pass, wouldn't it be better suited in `llvm/lib/Codegen` like other IR codegen passes?

topperc commented on llvm/include/llvm/Transforms/Vectorize/EVLIndVarSimplify.h:
```diff
@@ -0,0 +1,34 @@
+//===-------- EVLIndVarSimplify.h - Optimize vectorized loops w/ EVL IV----===//
```
Need `-*- C++ -*-` in headers I think

mshockwave replied:
Fixed.

topperc commented on llvm/test/CodeGen/RISCV/O3-pipeline.ll:
```diff
@@ -68,6 +68,9 @@
 ; CHECK-NEXT:       TLS Variable Hoist
+; CHECK-NEXT:       Scalar Evolution Analysis
+; CHECK-NEXT:       Loop Pass Manager
+; CHECK-NEXT:         EVL-based Induction Variables Simplify
```
Is possible to get this into the same loop pass manager as LSR?

mshockwave replied:
IIUC in order to do that we need to put EVLIndVarSimplify side-by-side with LSR, but I feel like we can make good uses of canonical form loops even between LSR and CodeGenPrepare. Do you think that will be the case?

topperc replied:
I'm not sure. LSR itself already does some things that make loops harder to analyze. Though it doesn't make them not countable.

mshockwave committed:
Move EVLIndVarSimplify to CodeGen

mshockwave commented:
> as this is a backend pass, wouldn't it be better suited in `llvm/lib/Codegen` like other IR codegen passes?

I agree, it's moved under CodeGen now.

nikolaypanchenko commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,254 @@
+#define DEBUG_TYPE "evl-iv-simplify"
+
+using namespace llvm;
+
```
Can you add an option to control that pass ?

mshockwave replied:
Done.

nikolaypanchenko commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,254 @@
+};
+} // anonymous namespace
+
+static std::optional<uint32_t> getVFFromIndVar(const SCEV *Step,
```
nit: `unsigned` as a return and `0` to mimic `nullopt` ?

mshockwave replied:
Good point, VF will never be zero. It's fixed now.

nikolaypanchenko commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,254 @@
+
+// Remove the original induction variable if it's not used anywhere.
+static void cleanupOriginalIndVar(PHINode *OrigIndVar, BasicBlock *InitBlock,
+                                  BasicBlock *BackEdgeBlock) {
```
Why can't you use `OrigIndVar` to get incoming BBs and values ? What if passed BBs are invalid for the phi ?

mshockwave replied:
It's fixed now. I use InductionDescriptor to obtain the initial value.

nikolaypanchenko commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,254 @@
+
+  // If the only user of OrigIndVar is the one produces RecValue, then we can
+  // safely remove it.
+  if (!OrigIndVar->hasOneUse() || OrigIndVar->user_back() != RecValue)
```
nit: presence of the checks makes me think that function should be named `tryToCleanupOriginalIndVar`

nikolaypanchenko replied:
also, do you need to check number of incoming values ?

mshockwave replied:
I've renamed the function as well as check the number of incoming values.

nikolaypanchenko commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,254 @@
+  BasicBlock *BB = IndVar->getParent();
+
+  Value *EVLIndex = nullptr;
+  Value *RemVL = nullptr, *AVL = nullptr;
```
to be in sync with vectorizer, better to `RemVL` -> `RemTC`, `AVL` -> `TC`

nikolaypanchenko commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,254 @@
+  using namespace PatternMatch;
+  BasicBlock *BB = IndVar->getParent();
+
+  Value *EVLIndex = nullptr;
```
just `EVL` or `EVLIndVar`

mshockwave replied:
Done.

mshockwave committed:
Address review comments

nikolaypanchenko reviewed (commented):
lgtm

topperc commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,266 @@
+                    "EVL-based Induction Variables Simplify", false, false)
+
+bool EVLIndVarSimplify::runOnLoop(Loop *L, LPPassManager &LPM) {
+  auto &SE = getAnalysis<ScalarEvolutionWrapperPass>().getSE();
```
Do we need to check skipLoop here for opt-bisect?

mshockwave replied:
Done.

topperc commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,266 @@
+      ConstantRange CR = llvm::getVScaleRange(&F, 64);
+      if (const APInt *Fixed = CR.getSingleElement()) {
+        V = V.zextOrTrunc(Fixed->getBitWidth());
+        uint64_t VF = V.udiv(*Fixed).getLimitedValue();
```
Is it possible the constant isn't evenly divisible by vscale?

mshockwave replied:
I added a check here.

topperc commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,266 @@
+
+  if (!EVLIndVar || !TC)
+    return false;
+
```
Do we need to check that `TC` is related to the original trip count calculation? Couldn't we construct an invalid loop that has a different trip count for the vsetvli than the countable trip count?

mshockwave replied:
Ideally we can do this by comparing the TC's SCEV with canonical trip count's SCEV. But it turns out canonical trip count is not always SCEV-computable. So I only do such comparison when both TC and vscale are constants. Otherwise, I simply check if TC is used in the SCEV expressions of either the upper or lower bound of canonical IV.

topperc commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,266 @@
+  if (InitValue != IVD.getStartValue())
+    std::swap(InitValue, RecValue);
+
+  // If the only user of OrigIndVar is the one produces RecValue, then we can
```
"one produces" -> "one that produces"

mshockwave replied:
Fixed.

topperc commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,266 @@
+
+  LLVM_DEBUG(dbgs() << "Removed the original IndVar " << *OrigIndVar << "\n");
+  // Remove OrigIndVar by replacing all its uses by the initial value of this
+  // loop. Then DCE will take care of the rest.
```
There's no explicit DCE after this pass is there?

mshockwave replied:
No there isn't, though the dead code will be cleanup anyway later in the codegen pipeline. So I change the wording here to not mentioning DCE.

npanchen commented:
add @penzn

mshockwave committed:
Address review comments

And check if the trip count matches the canonical IV.

mshockwave committed:
fixup! Address review comments

preames commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,296 @@
+
+  LLVM_DEBUG(dbgs() << "Using " << *EVLIndVar << " for EVL-based IndVar\n");
+
+  // Create an EVL-based comparison and replace the branch to use it as
```
I want to suggest an alternate approach here.  It took me quite a while to try to understand what you're trying to prove here, and I think there's an easier and more general alternative.

First, terminology.  EVLStep = min(VF, sub_nsw(EVL_TC, IV))

Step 1 - Check that the step of the IV being replaced is the same as the EVL IV's step on all but the last iteration.  (You already do this, thought the comments don't explain the significance.)

Step 2 - Using SCEV's backedge taken count reasoning, compute the exit value for the EVL IV.

SCEV won't directly do this for you - that's the whole countability bit - but you should be able to compute it as:
IV->evaluateAtIteration(BTC) + EVLStep.evaluateAtIteration(BTC).  (You could simplify the EVLStep based on the final iteration assumption, but you don't need to, and avoiding it reduces your scope for bugs.)

Note that the fact we know EVLStep <= IV.Step is critical for the correctness of the above.  If we didn't, we'd have to prove non-trivial wrap properties.  

Step 3 - Rewrite the exit condition to be an equality comparison between the EVLIV and the new exit value.  Note that you don't need the equality test restriction from the current code.  An important point is that the exit value above is very likely to be EVL_TC in practice, but that proving that is a proof burden you don't actually need.

Use SCEVExpander for this, and let it worry about all the simplifications.

mshockwave replied:
Thank you for the suggestion, I think `IV->evaluateAtIteration(BTC) + EVLStep.evaluateAtIteration(BTC)` is pretty neat.

I have one question here though:
> First, terminology. EVLStep = min(VF, sub_nsw(EVL_TC, IV))

Is it supposed to be EVLStep = min(_IV.Step_, sub_nsw(EVL_TC, IV))?

preames replied:
No, it wasn't.  The active_vector_length has the VF parameter.  That *may* be the same as the step, but in principle doesn't have to be.  Note that my (1) above (which your code already does) is proving VF=IV.Step, but that is a step you need to prove.  It's not something you can assume.

mshockwave replied:
I think there is another problem here: the backedge taken count cannot be computed when vscale is not constant. Namely, IV.Step is in the most commonly seen form of `(VF x vscale)`. Which also causes a problem in another comment thread where you suggested avoid using `getBounds`, because I guess you were suggesting to get the final IV SCEV via something like `IV.evaluateAtIteration(CanonicalTripCount)`, but CanonicalTripCount, just like backedge taken count, cannot be computed in this (i.e. non-constant vscale) case.

However, alternatively I think we can get exit value via this:
```
IV_BTC = IV_Final - IV.Step
Tail = TC - IV_BTC
Exit value = IV_BTC + Tail
```
`IV_BTC` as the name suggested, is the equivalence of `IV->evaluateAtIteration(BTC)` you suggested, but this time we calculate it from `IV_Final`, which is, well, final IV value returned from `getBounds`. Since it only uses IR pattern matching we don't have to worry about the aforementioned uncomputable problem.

preames replied:
> think there is another problem here: the backedge taken count cannot be computed when vscale is not constant. Namely, IV.Step is in the most commonly seen form of `(VF x vscale)`....  just like backedge taken count, cannot be computed in this (i.e. non-constant vscale) case.

Ok, this seems weird.  I don't know of any fundamental reason that SCEV shouldn't be able to figure out a trip count for a loop invariant step - even if non constant.  Would you mind sharing a piece of IR which demonstrates this?  I want to see if this is something easy to fix, or I'm forgetting some subtle interaction.

mshockwave replied:
> Would you mind sharing a piece of IR which demonstrates this? I want to see if this is something easy to fix, or I'm forgetting some subtle interaction.

I was using the first function (i.e. `@simple`) in the test for this PR (`evl-iv-simplify.ll`). I actually traced down to why it could not compute: the smoking gun is `ScalarEvolution::howFarToZero`, specifically, [this line](https://github.com/llvm/llvm-project/blob/7dc84e225e11e37925db6f4f08269f447d2f2347/llvm/lib/Analysis/ScalarEvolution.cpp#L10494) that actually bails out with SCEVCouldNotCompute. As the TODO comment right above it suggested, it couldn't handle non-constant step in AddRec for now.

preames replied:
> > Would you mind sharing a piece of IR which demonstrates this? I want to see if this is something easy to fix, or I'm forgetting some subtle interaction.
> 
> I was using the first function (i.e. `@simple`) in the test for this PR (`evl-iv-simplify.ll`). I actually traced down to why it could not compute: the smoking gun is `ScalarEvolution::howFarToZero`, specifically, [this line](https://github.com/llvm/llvm-project/blob/7dc84e225e11e37925db6f4f08269f447d2f2347/llvm/lib/Analysis/ScalarEvolution.cpp#L10494) that actually bails out with SCEVCouldNotCompute. As the TODO comment right above it suggested, it couldn't handle non-constant step in AddRec for now.

Thanks!

That limitation actually looks pretty easy to at least partially remove.  I'm currently a bit buried, but expect a patch in the next few days.

preames replied:
See https://github.com/llvm/llvm-project/pull/94411

Though this is with two changes to your test:
* adding vscale_range to the function
* making the IV nuw

I think both should apply to your original (i.e. your test is slightly over-reduced), but please confirm.

mshockwave replied:
> See #94411
> 
> Though this is with two changes to your test:
> 
> * adding vscale_range to the function
> * making the IV nuw
> 
> I think both should apply to your original (i.e. your test is slightly over-reduced), but please confirm.

Thank you for the quick fix.
I gave it a try and found 1 minor and 1 major concern: the minor one is that `vscale_range` has to be set otherwise BTC cannot be computed. This is because when we're computing BTC,  one of the preconditions is that the range of `(VF x vscale)` has to be non-zero, which depends on the range of vscale. Currently we use `[1,0)` ,a.k.a wrap around starting from 1, to represent the range of vscale when `vscale_range` is not set -- but this is an _unsigned_ wrap around. However, when computing the multiplication range of `(VF x vscale)` this `[1,0)` is treated as _signed_ wrap around, leading to `(VF x vscale)`'s range being full_set. Therefore when being asked the minimum unsigned value of `(VF x vscale)`'s range, it returns zero. I call this "minor" because I believe in practice vscale_range will mostly be set. But it's still a little annoyed to be that we cannot represent explicit unsigned range like `[1, UINT_MAX]` with `ConstantRange`.

The major concern being that the instructions for computing exit value expanded from the SCEV expressions we crafted is pretty verbose, long, a more importantly, contain division instructions. Let me start with the formula we have:
```
IV = the original canonical IV
EVL_TC = the trip count in the scalar loop
IV.Step = (VF x vscale)
EVLStep = umin(IV.Step, sub_nsw(EVL_TC, IV))
BTC = backedge taken count in the original canonical IV loop
ExitValue = IV.evaluateAtIteration(BTC) + EVLStep.evaluateAtIteration(BTC)
```
In your original formula EVLStep is `min(VF, sub_nsw(EVL_TC, IV))`, which I had a question on whether the first operand of `min` is VF or `IV.Step`. Because to my best understandings, `VF` is the same VF we put in the scalable vector type we use, `(vscale x VF x ...)`, and also the VF used in computing step value in IV: `(VF x vscale)`. So I think we should use IV.Step here. For the very least, if I use `VF` (as in my definition) here, the result will be pretty wrong.

Now, since only SCEVAddRec has `evaluateAtIteration`, we can't really do `EVLStep.evaluateAtIteration`. Instead, I rewrote `EVLStep.evaluateAtIteration` into:
```
umin(IV.Step, sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC))
```
Because IV.Step and EVL_TC are both loop invariant.
Using the `@simple` function in my test case + the same modifications you made in your comment above, as an example, this final `IV.evaluateAtIteration(BTC) + umin(IV.Step, sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC))` looks like:
```
((4 * vscale * ((-1 * vscale * (4 + (-4 * ((-1 + (4 * vscale)<nuw><nsw> + %N) /u (4 * vscale)<nuw><nsw>))<nsw>)<nsw>) /u (4 * vscale)<nuw><nsw>)) + (((-4 * vscale * ((-1 * vscale * (4 + (-4 * ((-1 + (4 * vscale)<nuw><nsw> + %N) /u (4 * vscale)<nuw><nsw>))<nsw>)<nsw>) /u (4 * vscale)<nuw><nsw>)) + %N) umin (4 * vscale)<nuw><nsw>))
```
And those divisions seen in the expressions will be expanded into division instructions. I'm concern whether we should spend so many instructions to compute something that will almost certain be `EVL_TC`.

I also run the same algorithm on the `@fixed_iv_step_tc` function in my test case, which has fixed IV.Step and fixed trip count. The generated exit value is correct, which means my methodology is probably not terribly off.

I've also tried to use `sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC)`, which is the tail in the last iteration, in replacement of `umin(IV.Step, sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC))`. But the resulting Exit value, `IV.evaluateAtIteration(BTC) + sub_nsw(EVL_TC, IV.evaluateAtIteration(BTC)` is not really meaningful (it will certainly be `EVL_TC` though).

Last but not the least, I'm still a little confused about how to do Step (1) you described: it's true that I already did such a check -- but only on cases where both trip count and vscale are constant. For every other cases we only have SCEV expressions, which we cannot know it value during compile time. Even we expand the check into runtime check, what should we do if the check fails during runtime? Do we fall back the original canonical IV?

preames replied:
> > See #94411
> > Though this is with two changes to your test:
> > 
> > * adding vscale_range to the function
> > * making the IV nuw
> > 
> > I think both should apply to your original (i.e. your test is slightly over-reduced), but please confirm.
> 
> Thank you for the quick fix. I gave it a try and found 1 minor and 1 major concern: the minor one is that `vscale_range` has to be set otherwise BTC cannot be computed. This is because when we're computing BTC, one of the preconditions is that the range of `(VF x vscale)` has to be non-zero, which depends on the range of vscale. Currently we use `[1,0)` ,a.k.a wrap around starting from 1, to represent the range of vscale when `vscale_range` is not set -- but this is an _unsigned_ wrap around. However, when computing the multiplication range of `(VF x vscale)` this `[1,0)` is treated as _signed_ wrap around, leading to `(VF x vscale)`'s range being full_set. Therefore when being asked the minimum unsigned value of `(VF x vscale)`'s range, it returns zero. I call this "minor" because I believe in practice vscale_range will mostly be set. But it's still a little annoyed to be that we cannot represent explicit unsigned range like `[1, UINT_MAX]` with `ConstantRange`.

Clang emits the vscale_range by default.  vscale is bounded by the values of possible VLEN, and VLEN has a specific upper bound in the RISCV vector specification.  So while annoying that SCEV can't just ask the backend for this, this is a non-problem for any real c/c++ input.  

For the rest of your response, one quick comment - the divisions are by 4 * vscale.  We know that vscale is a power of two by definition, so these are right shifts.  You do need a popcount to know which power of two, but we could reasonable emit one popcount and a bunch of shifts.  (Not saying we do so today - I'm saying it's possible.)

For the rest, I feel we are talking past each other - and frankly, I'm having trouble following the detail in github PR interface - can I suggest we setup a brief phone call to chat this through instead?

mshockwave replied:
> For the rest, I feel we are talking past each other - and frankly, I'm having trouble following the detail in github PR interface - can I suggest we setup a brief phone call to chat this through instead?

That will be really helpful! I'll send you an email to your "public" email address to coordinate this.

mshockwave committed:
Use RecursivelyDeleteDeadPHINode instead of cleaning IndVar manually

mshockwave commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,296 @@
+}
+
+// Remove the original induction variable if it's not used anywhere.
+static void tryCleanupOriginalIndVar(PHINode *OrigIndVar,
```
Good point, it's fixed now.

preames commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,296 @@
+  }
+
+  // Retrieve the loop bounds.
+  std::optional<Loop::LoopBounds> Bounds = L.getBounds(SE);
```
IVDescriptors has limited use of SCEV to identify addrecs.  LoopBounds::getBounds works solely on IR, matching existing instructions.  It doesn't produce the limit or start in terms of SCEV values.

mshockwave committed:
Use a better exit condition that is derived from the original loop

The new exit condition will not be `EVLIV > VF * BTC` where VF is the
vectorization factor and BTC being the backedge taken count.

mshockwave commented:
I'd just updated this PR to use a new exit condition. Previously, the exit condition of the new EVL-based loop was implemented by comparing IV against a trip count retrieved from IR pattern matching. This turned out to be risky since the said condition was not derived from the original canonical loop, so there is a chance that the number of iterations of the new EVL-based loop might not match that of the original loop.

The new exit condition is `EVLIV > VF * BTC`, where `EVLIV` is the new induction variable, `VF` is the vectorization factor and `BTC` is backedge taken count (i.e. trip count minus one in most cases). I think I have a proof written somewhere showing that this new condition is sound. I'll share the proof here once I found it. It is worth noting that this new exit condition is based upon the new semantics of `llvm.experimental.get.vector.length` landed in https://github.com/llvm/llvm-project/commit/e8063702cfbbf39f0b92283d0588dee264b5eb2b.

I've verified the correctness of the new exit condition with SPEC2006INT and SPEC2017INTRATE. In these two suites, the dynamic instruction counts keep mostly the same, with minor (<1%) improvements in some benchmarks.

One thing which I haven't had a good solution yet is the fact that SCEVExpander (or just ScalarEvolution in general) couldn't simplify the SCEV expression for `VF * BTC` in some cases. For instance, the `@simple` function in `test/CodeGen/RISCV/evl-iv-simplify.ll` uses a bunch of division and multiplication to build the said expression. Though in cases where `vscale_range` are singular ScalarEvolution successfully simplifies the expression. Plus, my experiments also showed that the dynamic instruction counts barely increase. @preames what do you think on this particular issue?

wangpc-pp commented:
> The new exit condition is EVLIV > VF * BTC, where EVLIV is the new induction variable, VF is the vectorization factor and BTC is backedge taken count (i.e. trip count minus one in most cases).

I don't know if I understand correctly, maybe https://github.com/llvm/llvm-project/pull/102575 is related to this change?

mshockwave commented:
> > The new exit condition is EVLIV > VF * BTC, where EVLIV is the new induction variable, VF is the vectorization factor and BTC is backedge taken count (i.e. trip count minus one in most cases).
> 
> I don't know if I understand correctly, maybe #102575 is related to this change?

thanks, I tried your PR but unfortunately it didn't help a lot...

mshockwave commented:
ping

mshockwave committed:
Merge branch 'main' into patch/evl-vectorized-iv-simplified

mshockwave committed:
Merge branch 'main' into patch/evl-vectorized-iv-simplified

mshockwave committed:
Merge branch 'main' into patch/evl-vectorized-iv-simplified

mshockwave committed:
(Stash) Put EVLIVSimplify Pass at the end of the vectorizer Pass

In order to simplify the exit condition.

mshockwave committed:
Merge branch 'main' into patch/evl-vectorized-iv-simplified

arcbbb commented:
Related pull request: https://github.com/llvm/llvm-project/pull/117506, addressing the removal of the original canonical IV.

mshockwave commented on llvm/lib/CodeGen/EVLIndVarSimplify.cpp:
```diff
@@ -0,0 +1,271 @@
+    // Check 3: Pattern match to find the EVL-based index.
+    if (match(RecValue,
+              m_c_Add(m_ZExtOrSelf(IntrinsicMatch), m_Specific(&PN))) &&
+        match(RemTC, m_Sub(m_Value(), m_Specific(&PN)))) {
```
> having the vectorizer do this transform during codegen. It knows by construction the two BTCs are equal

I actually have been testing an alternative solution that using IR pattern matching to find the original trip count (similar to the very first version of this PR) but putting the Pass right after the LoopVectorizer. Based on the assumption that loop vectorizer will produce the desired "shape" of loops that we can easily extract the trip count / BTC without fear. But I guess even with this, we somehow need to make sure the LV produces the same BTCs for the two IVs just like you mentioned.

mshockwave replied:
> Consider an EVL IV where the sub is M, and the canonical IV where BTC is in terms of N. N != M. In this case, the original loop would run a different number of iterations than the modified loop unless you can prove N and M are the same

To my best understandings, due to how LoopVectorizer derives N and M from the original trip count, N and M will be the same in the code produced by LV. So with my alternative solution which run this Pass right after LV, I think we can assure that N and M as seen by this Pass will be the same.

mshockwave committed:
Merge branch 'main' into patch/evl-vectorized-iv-simplified

mshockwave committed:
Update tests

mshockwave committed:
Removed legacy Pass

Since we're not adding to the codegen pipeline anymore

mshockwave commented:
I've pushed the latest version of this Pass that relies on IR pattern matching against the specific shape of loops generated by LoopVectorizer. Which means that this Pass now runs right after the LV.

mshockwave committed:
Update tests

mshockwave commented:
Closes in favor of #131005
